### ISSUE TITLE
Replace deprecated/discouraged `XML::Simple` with `XML::Twig`

### ISSUE THREAD

**Issue by jonasbn** [jonasbn/perl-workflow#126]

**Comment by jonasbn** [comment]
What is the risk is regard to backwards compatibility? could this be done prior to version 2.0 ?

**Comment by pastobas1966** [comment]
I've suggested using Workflow in my current job. One of our lead developers expressed concern wrt XML::Simple.

Looks like YAML is an alternative option but I wondered if there's been any progress with this ticket? Also, I wondered why XML::Twig in preference to other XML modules?

**Comment by ehuelsmann** [comment]
> I've suggested using Workflow in my current job. One of our lead developers expressed concern wrt XML::Simple.

@pastobas1966 thanks for posting your feedback. Although it would indeed be nice and a good step forward to implement other XML parsers than XML::Simple, it's hard to maintain backward compatibility while doing that replacement.

Did your lead express any specifics with respect to the use of XML::Simple? We may be able to verify if those apply to our code base.

**Comment by ehuelsmann** [comment]
@jonasbn I don't think this is possible without breaking existing XML configurations. The reason being that XML::Simple maps totally different things to the same output; which XML::Twig doesn't. In XML::Simple, the following are the same (but not in XML::Twig):

```xml
<workflow type="myworkflow" class="My::Class" initial_state="MY_INITIAL">
</workflow>
```

```xml
<workflow>
    <type>myworkflow</type>
   <class>My::Class</class>
   <initial_state>MY_INITIAL</initial_state>
</workflow>
```

Both are allowable by our current XML::Simple use; they have totally different XSD schemas.

**Comment by ehuelsmann** [comment]
I think OpenXPKI doesn't use XML config. LedgerSMB does; I'm happy to make the switch from one to the other, but I'd prefer to have a schema released *now* and make the switch to enforcing it in one or two years.

**Comment by jonasbn** [comment]
@ehuelsmann sounds reasonable

**Comment by ehuelsmann** [comment]
@pastobas1966 Workflow 2.0 comes with a YAML version for the config too. So you don't need to use XML or Perl config anymore. Please check it out and let us know if your management still has the same issues.
